demos/guide/tls-client-block.c

```diff
@@ -0,0 +1,252 @@
+     * "socket" here for this and the subsequent lookup, connect and close
+     * functions. But for portability reasons and also so that we get errors
+     * on the OpenSSL stack in the event of a failure we use OpenSSL's
+     * versions of these functions.
+     */
+    sock = BIO_socket(AF_INET, SOCK_STREAM, 0, 0);
+    if (sock == -1)
+        return NULL;
+
+    /*
+     * Lookup IP address info for the server. For simplicity we only look at
+     * the first address returned.
+     */
+    if (!BIO_lookup_ex(hostname, port, BIO_LOOKUP_CLIENT, AF_INET,
+                       SOCK_STREAM, 0, &ai)) {
+        BIO_closesocket(sock);
+        return NULL;
+    }
+
+    /* Connect the socket to the server's address */
+    if (!BIO_connect(sock, BIO_ADDRINFO_address(ai), BIO_SOCK_NODELAY)) {
+        BIO_closesocket(sock);
+        BIO_ADDRINFO_free(ai);
+        return NULL;
+    }
+
+    /* Free the address information resources we allocated earlier */
+    BIO_ADDRINFO_free(ai);
+
+    /* Create a BIO to wrap the socket*/
+    bio = BIO_new(BIO_s_socket());
+    if (bio == NULL)
+        BIO_closesocket(sock);
+
+    /*
+     * Associate the newly created BIO with the underlying socket. By
+     * passing BIO_CLOSE here the socket will be automatically closed when
+     * the BIO is freed. Alternatively you can use BIO_NOCLOSE, in which
+     * case you must close the socket explicitly when it is no longer
+     * needed.
+     */
+    BIO_set_fd(bio, sock, BIO_CLOSE);
+
+    return bio;
+}
+
+/* Server hostname and port details */
+#define HOSTNAME "www.example.com"
+#define PORT     "443"
+
+/*
+ * Simple application to send a basic HTTP/1.0 request to a server and
+ * print the response on the screen.
+ */
+int main(void)
+{
+    SSL_CTX *ctx = NULL;
+    SSL *ssl;
+    BIO *bio = NULL;
+    int res = EXIT_FAILURE;
+    int ret;
+    const char *request = "GET / HTTP/1.0\r\nHost: "HOSTNAME"\r\n\r\n";
+    size_t written, readbytes;
+    char buf[160];
+
+    /*
+     * Create an SSL_CTX which we can use to create SSL objects from. We
+     * want an SSL_CTX for creating clients so we use TLS_client_method()
+     * here.
+     */
+    ctx = SSL_CTX_new(TLS_client_method());
+    if (ctx == NULL) {
+        printf("Failed to create the SSL_CTX\n");
+        goto end;
+    }
+
+    /*
+     * Configure the client to abort the handshake if certificate
+     * verification fails. Virtually all clients should do this unless you
+     * really know what you are doing.
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+
+    /* Use the default trusted certificate store */
+    if (!SSL_CTX_set_default_verify_paths(ctx)) {
+        printf("Failed to set the default trusted certificate store\n");
+        goto end;
+    }
+
+    /* Create an SSL object to represent the TLS connection */
+    ssl = SSL_new(ctx);
+    if (ssl == NULL) {
+        printf("Failed to create the SSL object\n");
```

Comment:
Call same wrapper to dump the error stack and then print the message. Here and below as appropriate.